articles/tutorials/qft-tutorial.md

```diff
@@ -0,0 +1,624 @@
+```
+
+The `mutable` keyword prefacing `resultArray` allows the variable to be rebound later in the code---for example, when adding our measurement results.
+
+#### Perform measurements in a `for` loop and add results to array
+
+After the Fourier transorm operations inside the `using` block, insert the following code:
+
+```qsharp
+            for(i in IndexRange(qs)) {
+				set resultArray w/= i <- M(qs[i]);
+            }
+```
+The [`IndexRange`](xref:microsoft.quantum.arrays.indexrange) function called on an array (e.g. our array of qubits, `qs`) returns a range over the indices of the array. 
+Here, we use it in our `for` loop to sequentially measure each qubit using the `M(qs[i])` statement.
+Each measured `Result` type (either `Zero` or `One`) is then added to the corresponding index position in `resultArray` with an update-and-reassign statement.
+
+> [!NOTE]
+> The syntax of this statement is unique to Q#, but corresponds to the similar `set r[i] <- M(qs[i]);` seen in other languages.
+
+The keyword `set` is always used to reassign variables bound using `mutable`.
+
+#### Return `resultArray`
+
+With all three qubits measured and the results added to `resultArray`, we are safe to reset and de-allocate the qubits as before.
+After the `using` block's close, insert
+
+```qsharp
+        return resultArray;
+```
+to ultimately return the output of our operation. 
+
+### Understanding the effects of measurement
+
+Let's change the placement of our `DumpMachine` functions to output the state before and after the measurements.
+The final operation code should look like: 
+
+```qsharp
+    operation Perform3QubitQFT() : Result[] {
+
+        mutable r = new Result[3];
+
+        using (qs = Qubit[3]) {
+
+            //QFT:
+            //first qubit:
+            H(qs[0]);
+            Controlled R1([qs[1]], (PI()/2.0, qs[0]));
+            Controlled R1([qs[2]], (PI()/4.0, qs[0]));
+
+            //second qubit:
+            H(qs[1]);
+            Controlled R1([qs[2]], (PI()/2.0, qs[1]));
+
+            //third qubit:
+            H(qs[2]);
+
+            SWAP(qs[2], qs[0]);
+
+            Message("Before measurement: ");
+            DumpMachine();
+
+            for(i in IndexRange(qs)) {
+                set r w/= i <- M(qs[i]);
+            }
+
+            Message("After measurement: ");
+            DumpMachine();
+
+            ResetAll(qs);
+        }
+        return r;
+    }
+}
+```
+
+Additionally, update your host file to process the returned array:
+
+#### [Python](#tab/tabid-python)
+```python
+import qsharp
+from Quantum.OperationsQFT import Perform3QubitQFT
+
+measurementResult = Perform3QubitQFT.simulate()
+print("\n")
+print("Measured post-QFT state: [qubit0, qubit1, qubit2]")
+print(measurementResult)
+
+# reversing order to show corresponding basis state in binary form
+binaryCompBasisState = ""
+for i in measurementResult:
+    binaryCompBasisState = str(i) + binaryCompBasisState
```

Comment:
Why do you need to reverse the order here? It seems weird to print `[1,1,0]` as `|011>`; I would have expected  `|110>`. Alternatively, if the problem is that the Q# code always returns its results in the wrong order, why not fix up the order on the Q# side?

Separately, you _could_ write this loop as a one-liner

    binaryCompBasisState = ''.join(str(i) for i in reversed(measurementResult))

or (less good IMHO)

    binaryCompBasisState = ''.join(map(str, reversed(measurementResult)))

which might eliminate the need for any explicit comment about "reversing order." (And if you decide you don't need to reverse anything, then the code change is obvious: just drop the word `reversed` from the one-liner. Obvious code is good code.)

Reply:
Unfortunately the problem is more in that there's not necessarily a "right" order here. It's a slight annoyance that comes from the options we have to represent the computational basis states of a register of qubits: either 
- "big-endian", where the left-most bit corresponds to the first qubit in the register (the `[1,1,0]` here), or
- "little-endian", where that first qubit in the register corresponds to the right-most bit.
The big-endian format is particularly amenable to viewing these operations/states in the linear algebra context, so it's the norm in a lot of quantum computing literature. But, as seen here, representing those basis states with integers instead of bit strings is most useful with little-endian: 
- state `|0> being all qubits in `|0>`: `|0...00>`, 
- `|1>` having the _first_ qubit in the register as `|1>`:  `|0...01>`
- `|2>`: `|0...10>`, and so on.

So, the goal here was to be as clear as possible with how the measurement results of the first, second, and third qubits translate to the integers they're seeing printed by the dump function. Will add more explanation!